Fix missing initializations for models created in 2023

bvantuan opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes missing weight initializations for models created in 2023.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@Cyrilvallez 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @zach-huggingface, @SunMarc and @qgallouedec
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc @zach-huggingface
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

bvantuan committed:
fix SwiftFormer

bvantuan committed:
fix Kosmos2

bvantuan committed:
fix Owlv2

bvantuan committed:
fix Sam

bvantuan committed:
fix Vits

bvantuan committed:
fix Pvt

bvantuan committed:
fix MobileViTV2

bvantuan committed:
fix PatchTST

bvantuan committed:
fix Bros

bvantuan committed:
fix Informer

Cyrilvallez commented:
Super happy to see you're following again on a new batch of models @bvantuan! 🚀🤗 Let me know when this is ready!

bvantuan committed:
fix BridgeTower

bvantuan committed:
fix Mra and Yoso

bvantuan committed:
fix Rwkv

bvantuan committed:
fix EfficientNet

bvantuan committed:
fix NllbMoe

bvantuan committed:
fix Tvp

bvantuan committed:
fix Clap

bvantuan committed:
fix Autoformer

bvantuan committed:
fix SwiftFormer

bvantuan committed:
fix Mgpstr

bvantuan committed:
fix Align

bvantuan committed:
fix VitMatte

bvantuan committed:
fix SpeechT5

bvantuan committed:
add conditional check for parameters

bvantuan committed:
fix SpeechT5

bvantuan committed:
fix TimmBackbone and Clvp

bvantuan committed:
fix SwiftFormer

bvantuan committed:
fix SeamlessM4T and SeamlessM4Tv2

bvantuan committed:
fix Align

bvantuan committed:
fix Owlv2 and OwlViT

bvantuan commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -441,9 +440,28 @@ def _init_weights(self, module):
-                module.time_mix_receptance.data = torch.pow(time_weight, ratio_1_to_almost0)
+            module.time_mix_key.data = torch.pow(time_weight, ratio_1_to_almost0)
+            module.time_mix_receptance.data = torch.pow(time_weight, ratio_1_to_almost0)
+        elif isinstance(module, nn.Linear):
+            shape = module.weight.data.shape
+            gain = 1.0
+            scale = 1.0  # extra scale for gain
+            if module.bias is not None:
+                module.bias.data.zero_()
+            if shape[0] > shape[1]:
+                gain = math.sqrt(shape[0] / shape[1])
+            if shape[0] == self.config.vocab_size and shape[1] == self.config.hidden_size:  # final projection?
+                scale = 0.5
+
+            gain *= scale
+            nn.init.orthogonal_(module.weight, gain=gain)
+        elif isinstance(module, nn.Embedding):
+            shape = module.weight.data.shape
+            gain = 1e-4 * math.sqrt(max(shape[0], shape[1]))
+            nn.init.orthogonal_(module.weight, gain=gain)
```
https://github.com/BlinkDL/RWKV-LM/blob/cc2ee37c04422e37d1b022fbae60995bb4eebc2f/RWKV-v3/src/model.py#L62-L105

bvantuan commented:
Cc @Cyrilvallez ! The PR is now ready and awaiting your review😊.

Cyrilvallez reviewed (commented):
All right, nice work! 🤗 Very happy to see this! Just a few comments to simplify! 
Especially, the pattern `copy_(torch.randn(...))` is an anti-pattern: we should use `normal_` instead directly, which makes it much simpler and we don't have to worry about the shapes

bvantuan committed:
add reviewed changes

bvantuan committed:
Merge branch 'main' into fix-missing-initializations-2023

bvantuan committed:
add reviewed changes

bvantuan committed:
fix typo

bvantuan commented on src/transformers/models/bridgetower/modeling_bridgetower.py:
```diff
@@ -936,30 +936,30 @@ class BridgeTowerPreTrainedModel(PreTrainedModel):
+        if isinstance(module, BridgeTowerVisionTransformer):
+            proj_std = (module.transformer.hidden_size**-0.5) * ((2 * module.transformer.num_hidden_layers) ** -0.5)
+            attn_std = module.transformer.hidden_size**-0.5
+            fc_std = (2 * module.transformer.hidden_size) ** -0.5
```
Got it, that makes sense!

bvantuan commented on src/transformers/models/seamless_m4t/modeling_seamless_m4t.py:
```diff
@@ -2750,7 +2754,7 @@ def _reorder_cache(past_key_values, beam_idx):
 )
 class SeamlessM4TForSpeechToText(SeamlessM4TPreTrainedModel, GenerationMixin):
-    _keys_to_ignore_on_load_missing = ["text_decoder", "t2u_model", "vocoder"]
+    _keys_to_ignore_on_load_missing = ["t2u_model", "vocoder"]
```
The `SeamlessM4TForSpeechToText` class only needs the `speech_encoder` and `text_decoder`, so it's okay if the checkpoint being loaded also has weights for a `text_encoder`, `t2u_model`, or `vocoder`. It had a typo "text_decoder".

bvantuan commented on src/transformers/models/seamless_m4t_v2/modeling_seamless_m4t_v2.py:
```diff
@@ -2954,7 +2961,7 @@ def _reorder_cache(past_key_values, beam_idx):
 )
 class SeamlessM4Tv2ForSpeechToText(SeamlessM4Tv2PreTrainedModel, GenerationMixin):
-    _keys_to_ignore_on_load_missing = ["text_decoder", "t2u_model", "vocoder"]
+    _keys_to_ignore_on_load_missing = ["t2u_model", "vocoder"]
```
I explained as above

bvantuan commented:
Thanks a lot for reviewing, @Cyrilvallez ! I’m really glad you’re happy with it. I completely agree—using `normal_` instead of `copy_(torch.randn(...))` is much cleaner and more elegant. I’ve simplified that part and addressed the other comments as well.

bvantuan committed:
Merge branch 'main' into fix-missing-initializations-2023

bvantuan committed:
Merge branch 'main' into fix-missing-initializations-2023

bvantuan committed:
Merge branch 'main' into fix-missing-initializations-2023

Cyrilvallez committed:
Merge branch 'main' into fix-missing-initializations-2023

Cyrilvallez reviewed (approved):
Alright, thanks a lot for this new batch! Always greatly appreciated! 🤗

ydshieh commented:
Hi @bvantuan 

Thanks for the PR. Could you check

```
    "models_fastspeech2_conformer": {
        "job_link": "https://github.com/huggingface/transformers/actions/runs/16433389284/job/46439076372",
        "single": [
            {
                "line": "tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py::FastSpeech2ConformerModelTest::test_can_init_all_missing_weights",
                "trace": "(line 687)  AssertionError: False is not true : The following keys are not properly handled by `_init_weights()`:"
            },
            {
                "line": "tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py::FastSpeech2ConformerWithHifiGanTest::test_can_init_all_missing_weights",
                "trace": "(line 687)  AssertionError: False is not true : The following keys are not properly handled by `_init_weights()`:"
            }
        ]
    },
```
which are skipped before this PR but now failing since this PR?
